src/common/solaris_errno.cc

```diff
@@ -0,0 +1,499 @@
+#include "messages/MGetPoolStats.h"
+#include "messages/MGetPoolStatsReply.h"
+#include "messages/MHeartbeat.h"
+#include "messages/MInodeFileCaps.h"
+#include "messages/MLock.h"
+#include "messages/MLog.h"
+#include "messages/MLogAck.h"
+#include "messages/MMDSBeacon.h"
+#include "messages/MMDSCacheRejoin.h"
+#include "messages/MMDSFindIno.h"
+#include "messages/MMDSFindInoReply.h"
+#include "messages/MMDSFragmentNotify.h"
+#include "messages/MMDSLoadTargets.h"
+#include "messages/MMDSMap.h"
+#include "messages/MMDSOpenIno.h"
+#include "messages/MMDSOpenInoReply.h"
+#include "messages/MMDSResolve.h"
+#include "messages/MMDSResolveAck.h"
+#include "messages/MMDSSlaveRequest.h"
+#include "messages/MMDSTableRequest.h"
+#include "messages/MMonCommand.h"
+#include "messages/MMonCommandAck.h"
+#include "messages/MMonElection.h"
+#include "messages/MMonGetMap.h"
+#include "messages/MMonGetVersion.h"
+#include "messages/MMonGetVersionReply.h"
+#include "messages/MMonGlobalID.h"
+#include "messages/MMonHealth.h"
+#include "messages/MMonJoin.h"
+#include "messages/MMonMap.h"
+#include "messages/MMonPaxos.h"
+#include "messages/MMonProbe.h"
+#include "messages/MMonScrub.h"
+#include "messages/MMonSubscribe.h"
+#include "messages/MMonSubscribeAck.h"
+#include "messages/MMonSync.h"
+#include "messages/MOSDAlive.h"
+#include "messages/MOSDBoot.h"
+#include "messages/MOSDECSubOpRead.h"
+#include "messages/MOSDECSubOpReadReply.h"
+#include "messages/MOSDECSubOpWrite.h"
+#include "messages/MOSDECSubOpWriteReply.h"
+#include "messages/MOSDFailure.h"
+#include "messages/MOSDMap.h"
+#include "messages/MOSDMarkMeDown.h"
+#include "messages/MOSDOp.h"
+#include "messages/MOSDOpReply.h"
+#include "messages/MOSDPGBackfill.h"
+#include "messages/MOSDPGCreate.h"
+#include "messages/MOSDPGInfo.h"
+#include "messages/MOSDPGLog.h"
+#include "messages/MOSDPGMissing.h"
+#include "messages/MOSDPGNotify.h"
+#include "messages/MOSDPGPull.h"
+#include "messages/MOSDPGPush.h"
+#include "messages/MOSDPGPushReply.h"
+#include "messages/MOSDPGQuery.h"
+#include "messages/MOSDPGRemove.h"
+#include "messages/MOSDPGScan.h"
+#include "messages/MOSDPGTemp.h"
+#include "messages/MOSDPGTrim.h"
+#include "messages/MOSDPing.h"
+#include "messages/MOSDRepScrub.h"
+#include "messages/MOSDScrub.h"
+#include "messages/MOSDSubOp.h"
+#include "messages/MOSDSubOpReply.h"
+#include "messages/MPGStats.h"
+#include "messages/MPGStatsAck.h"
+#include "messages/MPing.h"
+#include "messages/MPoolOp.h"
+#include "messages/MPoolOpReply.h"
+#include "messages/MRecoveryReserve.h"
+#include "messages/MRemoveSnaps.h"
+#include "messages/MRoute.h"
+#include "messages/MStatfs.h"
+#include "messages/MStatfsReply.h"
+#include "messages/MTimeCheck.h"
+#include "messages/MWatchNotify.h"
+
+#define dout_subsys ceph_subsys_ms
+
+#include "solaris_errno.h"
+#include "solaris_message_errno.h"
+
+// converts from linux errno values to host values
+__s32 translate_errno(CephContext *cct,__s32 r) 
+{
+  if (r < -34) {
```

Comment:
I think a cleaner approach here would be to have a generic

 translate_errno()

method that is a #define on Linux but does this conversoin on solaris.  Then change all of the return sites in librados.cc to pass the return value through.  For example,  

https://github.com/ceph/ceph/blob/master/src/librados/librados.cc#L1149

would do

 return translate_errno(io_ctx_impl->sparse_read(oid, m, bl, len, off));